File: 152-maximum-product-subarray/152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
            currMin = 1
            currMax = 1
            answer = nums[0]
            for i in nums:
                tempArray = [i, i * currMin, i * currMax]
                currMax = max(tempArray)
                currMin = min(tempArray)
                
                answer = max(answer, currMax)
                
            return answer

File: 152-maximum-product-subarray/152-maximum-product-subarray.cpp
#include<algorithm>

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        
        // Approach - Dynamic Programming
        
        int curMax = 1;
        int curMin = 1;
        
        int result = nums[0]; 
        
        for (int i = 0; i < static_cast<int>(nums.size()); i++) {
            int currentNum = nums[i];
            
            std::vector<int> maxList = {currentNum, currentNum * curMin, currentNum * curMax};
            int maxFromList = *std::max_element(maxList.begin(), maxList.end());
            int minFromList = *std::min_element(maxList.begin(), maxList.end());
            
            curMax = maxFromList;
            curMin = minFromList;
            
            result = std::max(result, curMax);
        }
        
        return result;
    }
};